File: StableBinarySortWithO1Space.c
#include <stdio.h>
// Hàm hoán đổi hai phần tử
void hoanDoi(int* a, int* b) {
    int tam = *a;
    *a = *b; *b = tam;
}
void daoNguoc(int arr[], int l, int r) {
    if (l < r) {
        hoanDoi(&arr[l], &arr[r]);  // Hoán đổi arr[l] và arr[r]
        daoNguoc(arr, ++l, --r);     // Gọi đệ quy để tiếp tục hoán đổi các phần tử giữa
    }
}
void hopNhat(int arr[], int l, int m, int r) {
    int i = l;
    int j = m + 1;
    // Tìm chỉ số bắt đầu của phần tử dương trong nửa mảng bên trái (tìm chỉ số của Lp)
    while (i <= m && arr[i] < 0)
        i++;
    // Tìm chỉ số kết thúc của phần tử âm trong nửa mảng bên phải (tìm chỉ số của Rn)
    while (j <= r && arr[j] < 0)
        j++;
    j--;
    // Đảo ngược phần dương trong mảng con bên trái (arr[i..m])
    daoNguoc(arr, i, m);
    // Đảo ngược phần âm trong mảng con bên phải (arr[m+1..j])
    daoNguoc(arr, m + 1, j);
    // Đảo ngược lại phần arr[i..j]
    daoNguoc(arr, i, j);
}
// Hàm để phân chia dương và âm trong mảng
void sapXepPhanLoai(int arr[], int l, int r) {
    if (l < r) {
        int m = l + (r - l) / 2;  // Tính chỉ số giữa mảng
        // Gọi đệ quy để sắp xếp nửa đầu và nửa sau
        sapXepPhanLoai(arr, l, m);
        sapXepPhanLoai(arr, m + 1, r);

        hopNhat(arr, l, m, r);  // Hợp nhất mảng con
    }
}

int main() {
    int n; printf("Nhập số lượng phần tử của mảng: "); scanf("%d", &n);
    int arr[n]; printf("Nhập các phần tử của mảng: ");
    for (int i = 0; i < n; i++) {scanf("%d", &arr[i]);}
    sapXepPhanLoai(arr, 0, n - 1);
    printf("Mảng sau khi phân loại dương và âm: ");
    for (int i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }
    return 0;
}

File: StableBinarySortWithO1Space.py
def reverse(arr, l, r):
    if l < r:
        arr[l], arr[r] = arr[r], arr[l]  # Hoán đổi hai phần tử ở vị trí l và r
        reverse(arr, l + 1, r - 1)  # Gọi đệ quy để hoán đổi tiếp các phần tử giữa l và r
        
def merge(arr, l, m, r):
    i = l
    j = m + 1

    while i <= m and arr[i] < 0:  # Duyệt qua mảng từ l đến m để tìm phần tử dương đầu tiên
        i += 1

    while j <= r and arr[j] < 0:  # Duyệt qua mảng từ m+1 đến r để tìm phần tử âm cuối cùng
        j += 1
    j -= 1  # Điều chỉnh j để trỏ tới phần tử âm cuối cùng trong mảng con

    reverse(arr, i, m)  # Đảo ngược phần tử dương trong nửa mảng bên trái
    reverse(arr, m + 1, j)  # Đảo ngược phần tử âm trong nửa mảng bên phải
    reverse(arr, i, j)  # Đảo ngược lại phần từ i đến j để hoàn tất việc phân vùng âm-dương

def rearrange(arr, l, r):
    if l < r:  # Kiểm tra điều kiện dừng đệ quy
        m = l + (r - l) // 2  # Tính chỉ số m (trung điểm của mảng)

        rearrange(arr, l, m)
        rearrange(arr, m + 1, r)
        merge(arr, l, m, r)  # Hợp nhất hai mảng con sau khi phân chia

if __name__ == "__main__":
    # Yêu cầu nhập mảng từ bàn phím
    arr_input = input("Nhập mảng các số, cách nhau bằng dấu cách: ").split()
    
    # Chuyển đổi mảng nhập vào thành kiểu số nguyên
    arr = [int(num) for num in arr_input]
    
    # Gọi hàm rearrange với mảng đầu vào
    rearrange(arr, 0, len(arr) - 1)
    
    # In mảng sau khi phân chia
    print("Mảng sau khi phân chia âm và dương:")
    for ele in arr:
        print(ele, end=' ')  # In mảng sau khi phân chia
